### PR TITLE
feat: defer the database connection to when it's needed

### LINKED ISSUE #762: `WeaviateDocumentStore`: avoid connecting to the server in the class constructor

### PR DESCRIPTION
**Pull request by masci**
### Related Issues

- fixes #762 

### Checklist

- I have read the [contributors guidelines](https://github.com/deepset-ai/haystack-core-integrations/blob/main/CONTRIBUTING.md) and the [code of conduct](https://github.com/deepset-ai/haystack-core-integrations/blob/main/CODE_OF_CONDUCT.md)
- I have updated the related issue with new insights and changes
- I added unit tests and updated the docstrings
- I've used one of the [conventional commit types](https://www.conventionalcommits.org/en/v1.0.0/) for my PR title: `fix:`, `feat:`, `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`.

### CODE REVIEW

**Comment by masci** [comment]
To my joy, all the tests are green locally using the same Docker image https://tableflipper.com

File: integrations/weaviate/src/haystack_integrations/document_stores/weaviate/document_store.py [hunk 1]
```diff
@@ -139,49 +139,74 @@ def __init__(
...
+            additional_headers=self._additional_headers,
+            embedded_options=self._embedded_options,
             skip_init_checks=False,
         )
+
         self._client.connect()
 
         # Test connection, it will raise an exception if it fails.
         self._client.collections._get_all(simple=True)
+        self._client.collections._get_all(simple=True)
```
Comment (anakin87) [inline integrations/weaviate/src/haystack_integrations/document_stores/weaviate/document_store.py:195]: is this intentional?

**Review by anakin87 — APPROVED** [review]
Nice work!

File: integrations/weaviate/src/haystack_integrations/document_stores/weaviate/document_store.py (hunk 1 above)
Comment (masci) [reply integrations/weaviate/src/haystack_integrations/document_stores/weaviate/document_store.py:195]: No 🙈

**Pull request merged**